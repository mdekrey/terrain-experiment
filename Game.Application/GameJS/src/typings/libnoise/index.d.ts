declare module 'libnoise' {
    export namespace libnoise {
        export abstract class ModuleBase {
            get<T = number>(index: number): T;
            set<T = number>(index: number, value: T): T;
            length(): number;
            abstract getValue(x, y, z): number;
        }
        export enum QualityMode {
            LOW,
            MEDIUM,
            HIGH,
        }
        export namespace Utils {
            export function Clamp(value: number, min: number, max: number): number;
            export const Deg2Rad: number;
            export const GeneratorNoiseX: number;
            export const GeneratorNoiseY: number;
            export const GeneratorNoiseZ: number;
            export const GeneratorSeed: number;
            export const GeneratorShift: number;
            export function GradientCoherentNoise3D(x: number, y: number, z: number, seed: number, quality: QualityMode): number;
            export function GradientNoise3D(fx: number, fy: number, fz: number, ix: number, iy: number, iz: number, seed: number): number;
            export function InterpolateCubic(a: number, b: number, c: number, d: number, position: number): number;
            export function InterpolateLinear(a: number, b: number, position: number): number;
            export function MakeInt32Range(value: number): number;
            export function MakeCubicSCurve(value: number): number;
            export function MapQuinticSCurve(value: number): number;
            export const OctavesMaximum: number;
            export const Randoms: number[];
            export const SQRT3: number;
            export function ValueNoise3D(x: number, y: number, z: number, seed: number): number;
            export function ValueNoise3DInt(x: number, y: number, z: number, seed: number): number;
            export function fmod(a: number, b: number): number;
        }
        export namespace generator {
            export class Billow extends ModuleBase {
                constructor(frequency: number, lacunarity: number, persistence: number, octaves: number, seed: number, quality: QualityMode);
            }
            export class Checker extends ModuleBase {}
            export class Const extends ModuleBase {
                constructor(value: number);
            }
            export class Cylinder extends ModuleBase {
                constructor(frequency: number);
            }
            export class Perlin extends ModuleBase {
                constructor(frequency: number, lacunarity: number, persistence: number, octaves: number, seed: number, quality: QualityMode);
            }
            export class RidgedMultifractal extends ModuleBase {
                constructor(frequency: number, lacunarity: number, persistence: number, octaves: number, seed: number, quality: QualityMode);
            }
            export class Sphere extends ModuleBase {
                constructor(frequency: number);
            }
            export class Voronoi extends ModuleBase {
                constructor(frequency: number, displacement: number, seed: number, distance: number);
            }
        }
        export namespace operator {
            export class Abs extends ModuleBase {
                constructor(input: ModuleBase);
            }
            export class Add extends ModuleBase {
                constructor(lhs: number, rhs: number);
            }
            export class Blend extends ModuleBase {
                constructor(lhs: number, rhs: number, controller: number);
            }
            export class Cache extends ModuleBase {
                constructor(input: ModuleBase);
            }
            export class Clamp extends ModuleBase {
                constructor(min: number, max: number, input: ModuleBase);
            }
            export class Curve extends ModuleBase {
                constructor(input: ModuleBase);
            }
            export class Displace extends ModuleBase {
                constructor(input: ModuleBase, x: number, y: number, z);
            }
            export class Exponent extends ModuleBase {
                constructor(exponent: number, input: ModuleBase);
            }
            export class Invert extends ModuleBase {
                constructor(input: ModuleBase);
            }
            export class Max extends ModuleBase {
                constructor(lhs: number, rhs: number);
            }
            export class Min extends ModuleBase {
                constructor(lhs: number, rhs: number);
            }
            export class Multiply extends ModuleBase {
                constructor(lhs: number, rhs: number);
            }
            export class Power extends ModuleBase {
                constructor(lhs: number, rhs: number);
            }
            export class Rotate extends ModuleBase {
                constructor(rx: number, ry: number, rz: number, input: ModuleBase);
            }
            export class Scale extends ModuleBase {
                constructor(sx: number, sy: number, sz: number, input: ModuleBase);
            }
            export class ScaleBias extends ModuleBase {
                constructor(scale: number, bias: number, input: ModuleBase);
            }
            export class Select extends ModuleBase {
                constructor(min: number, max: number, fallOff: number, inputA: number, inputB: number, controller: number);
            }
            export class Subtract extends ModuleBase {
                constructor(lhs: number, rhs: number);
            }
            export class Terrace extends ModuleBase {
                constructor(input: ModuleBase, inverted: number);
            }
            export class Translate extends ModuleBase {
                constructor(x: number, y: number, z: number, input: ModuleBase);
            }
            export class Turbulence extends ModuleBase {
                constructor(power: number, input: ModuleBase, distortX: number, distortY: number, distortZ: number);
            }
        }
    }
}
