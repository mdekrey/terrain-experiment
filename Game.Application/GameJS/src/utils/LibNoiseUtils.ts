import { libnoise } from "libnoise";
import { construct, DataDrivenConstructor, DataDrivenInput, DataDrivenConstructorRecords } from "./DataDrivenComposition";

export const DEFAULT_PERLIN_LACUNARITY = 2;
export const DEFAULT_PERLIN_SEED = 0;
export const DEFAULT_PERLIN_OCTAVE_COUNT = 6;
export const DEFAULT_PERLIN_FREQUENCY = 1;
export const DEFAULT_PERLIN_PERSISTENCE = 0.5;

const yOffset = 1/400;

export function initializePerlin({
  lacunarity = DEFAULT_PERLIN_LACUNARITY,
  seed = DEFAULT_PERLIN_SEED,
}) {
  return dataDrivenNoise(perlinDataDrivenConstructor(seed, lacunarity));
}

export function perlinDataDrivenConstructor(seed: number, lacunarity: number): DataDrivenConstructor<number, libnoise.ModuleBase> {
  return {
    target: "operator.Translate",
    arguments: [0, yOffset, 0, {
      target: "operator.Clamp",
      arguments: [0, 1, {
        target: "operator.Add",
        arguments: [{
          target: "operator.Multiply",
          arguments: [{
            target: "generator.Perlin",
            arguments: [
              DEFAULT_PERLIN_FREQUENCY,
              lacunarity,
              DEFAULT_PERLIN_PERSISTENCE,
              DEFAULT_PERLIN_OCTAVE_COUNT,
              seed
            ]
          }, {
            target: "generator.Const",
            arguments: [1 / 1.77]
          }]
        }, {
          target: "generator.Const",
          arguments: [0.5]
        }]
      }]
    }]
  };
}

export function initializeRidgedMulti({
  seed = DEFAULT_PERLIN_SEED,
  scale = 1,
  lacunarity = DEFAULT_PERLIN_LACUNARITY
}) {
  return dataDrivenNoise(ridgedMultiDataDrivenConstructor(scale,  seed, lacunarity));
}

export function ridgedMultiDataDrivenConstructor(scale: number, seed: number, lacunarity: number): DataDrivenConstructor<number, libnoise.ModuleBase> {
  return {
    target: "operator.Translate",
    arguments: [0, yOffset, 0, {
      target: "operator.Scale",
      arguments: [scale, scale, 1, {
        target: "operator.Clamp",
        arguments: [0, 1, {
          target: "NormalRidge",
          arguments: [{
            target: "operator.Abs",
            arguments: [{
              target: "operator.Multiply",
              arguments: [{
                target: "generator.Perlin",
                arguments: [
                  DEFAULT_PERLIN_FREQUENCY,
                  lacunarity,
                  DEFAULT_PERLIN_PERSISTENCE,
                  DEFAULT_PERLIN_OCTAVE_COUNT,
                  seed
                ]
              }, {
                target: "generator.Const",
                arguments: [1 / 1.6]
              }]
            }]
          }]
        }]
      }]
    }]
  };
}

class Perlin extends libnoise.generator.Perlin {
  constructor(frequency: number, lacunarity: number, persistence: number, octaves: number, seed: number) {
    super(frequency, lacunarity, persistence, octaves, seed, libnoise.QualityMode.MEDIUM);
  }
}

class NormalRidge extends libnoise.ModuleBase {
  private base: libnoise.ModuleBase;
  constructor(base: libnoise.ModuleBase) {
    super();
    this.base = base;
  }
  getValue(x: number, y: number, z: number) {
    const val = libnoise.Utils.Clamp(this.base.getValue(x, y, z), 0, 1);
    return 1 - val * (1 - val) * 4;
  }
}

const dataDrivenConstructors: DataDrivenConstructorRecords<any, libnoise.ModuleBase> = {
  "NormalRidge": NormalRidge,
  "generator.Perlin": Perlin,
  "operator.Abs": libnoise.operator.Abs,
  "operator.Add": libnoise.operator.Add,
  "operator.Clamp": libnoise.operator.Clamp,
  "generator.Const": libnoise.generator.Const,
  "operator.Multiply": libnoise.operator.Multiply,
  "operator.Translate": libnoise.operator.Translate,
  "operator.Scale": libnoise.operator.Scale,
};

export function dataDrivenNoise<T extends DataDrivenInput<number, libnoise.ModuleBase>>(target: T) {
  return construct(target, dataDrivenConstructors);
}
