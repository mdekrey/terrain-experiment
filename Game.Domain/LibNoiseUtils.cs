using LibNoise;
using LibNoise.Generator;
using LibNoise.Operator;
using System;
using System.Collections.Generic;
using System.Text;

namespace Game.Domain
{
    public static class LibNoiseUtils
    {
        public const float DEFAULT_PERLIN_LACUNARITY = 2f;
        public const int DEFAULT_PERLIN_SEED = 0;
        public const int DEFAULT_PERLIN_OCTAVE_COUNT = 6;
        public const float DEFAULT_PERLIN_FREQUENCY = 1f;
        public const float DEFAULT_PERLIN_PERSISTENCE = 0.5f;

        public const float Y_OFFSET = 1 / 400f;

        public static ModuleBase InitializePerlin(float lacunarity = DEFAULT_PERLIN_LACUNARITY, int seed = 0, int octaveCount = DEFAULT_PERLIN_OCTAVE_COUNT, float frequency = DEFAULT_PERLIN_FREQUENCY)
        {
            return new Translate(0, Y_OFFSET, 0, 
                new Clamp(0, 1, 
                    new Add(
                        new Multiply(
                            new Perlin(frequency, lacunarity, DEFAULT_PERLIN_PERSISTENCE, octaveCount, seed, QualityMode.Medium), 
                            new Const(1 / 1.77f)
                        ), 
                        new Const(0.5f)
                    )
                )
            );
        }

        class NormalRidge : ModuleBase
        {
            private readonly ModuleBase original;

            public NormalRidge(ModuleBase original)
            {
                this.original = original;
            }

            public override double GetValue(double x, double y, double z)
            {
                var val = Utils.Clamp(original.GetValue(x, y, z), 0, 1);
                return 1 - val * (1 - val) * 4;
            }
        }
    }
}
