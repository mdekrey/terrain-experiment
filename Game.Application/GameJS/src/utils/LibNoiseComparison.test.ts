import { libnoise } from "libnoise";
import { DEFAULT_PERLIN_LACUNARITY, DEFAULT_PERLIN_PERSISTENCE, DEFAULT_PERLIN_OCTAVE_COUNT, DEFAULT_PERLIN_FREQUENCY, initializeRidgedMulti, initializePerlin } from "./LibNoiseUtils";

it("has the same perlin results as .net", () => {
    const perlin = new libnoise.generator.Perlin(DEFAULT_PERLIN_FREQUENCY, DEFAULT_PERLIN_LACUNARITY, DEFAULT_PERLIN_PERSISTENCE, DEFAULT_PERLIN_OCTAVE_COUNT,
        0, libnoise.QualityMode.MEDIUM);

    const step = 0.1;
    const microsteps = 30;

    const noise = Array.from(Array(microsteps).keys()).map(i => (i - microsteps / 2) / microsteps * step).map(v => perlin.getValue(v, 0, 0))
    expect(noise).toMatchSnapshot();
});

it("can normalize perlin results", () => {
    const ridged = initializePerlin({});

    const step = 0.1;
    const microsteps = 60;

    const noise = Array.from(Array(microsteps).keys()).map(i => (i - microsteps / 2) * step).map(v => ridged.getValue(v, 0, 0))
    expect(noise).toMatchSnapshot();
});

it("has the same ridged multifractal results as .net", () => {
    const ridged = new libnoise.generator.RidgedMultifractal(DEFAULT_PERLIN_FREQUENCY, 3.2, DEFAULT_PERLIN_OCTAVE_COUNT,
        0, libnoise.QualityMode.MEDIUM);

    const step = 1;
    const microsteps = 30;

    const noise = Array.from(Array(microsteps).keys()).map(i => (i - microsteps / 2) / microsteps * step).map(v => ridged.getValue(v, 0, 0))
    expect(noise).toMatchSnapshot();
});

it("can normalize ridged multifractal results", () => {
    const ridged = initializeRidgedMulti({});

    const step = 0.1;
    const microsteps = 60;

    const noise = Array.from(Array(microsteps).keys()).map(i => (i - microsteps / 2) * step).map(v => ridged.getValue(v, 0, 0))
    expect(noise).toMatchSnapshot();
});

it("changes ridged multifractal over large distances", () => {
    const ridged = new libnoise.generator.RidgedMultifractal(0.01, 3.2, DEFAULT_PERLIN_OCTAVE_COUNT,
        0, libnoise.QualityMode.MEDIUM);

    const step = 60;
    const microsteps = 60;

    const noise = Array.from(Array(microsteps).keys()).map(i => (i - microsteps / 2) * step).map(v => ridged.getValue(v, 0, 0))
    expect(noise).toMatchSnapshot();
});

it("has a reasonable range", () => {
    const perlin = new libnoise.generator.Perlin(DEFAULT_PERLIN_FREQUENCY, DEFAULT_PERLIN_LACUNARITY, DEFAULT_PERLIN_PERSISTENCE, DEFAULT_PERLIN_OCTAVE_COUNT,
        0, libnoise.QualityMode.MEDIUM);

    const step = 0.1;
    const microsteps = 200;

    const range = Array.from(Array(microsteps).keys())
        .map(i => i * step)
        .map(x => Array.from(Array(microsteps).keys())
        .map(i => i * step)
        .map(y => perlin.getValue(x, y, 0))
        .reduce((prev, next) => ({ min: Math.min(prev.min, next), max: Math.max(prev.max, next) }), { min: Number.POSITIVE_INFINITY, max: Number.NEGATIVE_INFINITY }))
        .reduce((prev, next) => ({ min: Math.min(prev.min, next.min), max: Math.max(prev.max, next.max) }), { min: Number.POSITIVE_INFINITY, max: Number.NEGATIVE_INFINITY });
    expect(range.min).toBeLessThan(-1.45);
    expect(range.min).toBeGreaterThan(-1.9);
    expect(range.max).toBeGreaterThan(1.45);
    expect(range.max).toBeLessThan(1.9);
});

it("has a ridged multi with a reasonable range", () => {
    const perlin = new libnoise.generator.RidgedMultifractal(DEFAULT_PERLIN_FREQUENCY, DEFAULT_PERLIN_LACUNARITY, DEFAULT_PERLIN_OCTAVE_COUNT,
        0, libnoise.QualityMode.MEDIUM);

    const step = 0.1;
    const microsteps = 200;

    const range = Array.from(Array(microsteps).keys())
        .map(i => i * step)
        .map(x => Array.from(Array(microsteps).keys())
        .map(i => i * step)
        .map(y => perlin.getValue(-x, -y, 0))
        .reduce((prev, next) => ({ min: Math.min(prev.min, next), max: Math.max(prev.max, next) }), { min: Number.POSITIVE_INFINITY, max: Number.NEGATIVE_INFINITY }))
        .reduce((prev, next) => ({ min: Math.min(prev.min, next.min), max: Math.max(prev.max, next.max) }), { min: Number.POSITIVE_INFINITY, max: Number.NEGATIVE_INFINITY });
    expect(range.min).toBeLessThan(-0.9);
    expect(range.min).toBeGreaterThan(-1);
    expect(range.max).toBeGreaterThan(1.45);
    expect(range.max).toBeLessThan(1.6);
});
