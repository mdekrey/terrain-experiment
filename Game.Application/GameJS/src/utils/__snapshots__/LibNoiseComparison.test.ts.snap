// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`has the same perlin results as .net 1`] = `
Array [
  0.15051893273369998,
  0.1358378323433078,
  0.12111284006069127,
  0.10714938081818756,
  0.09482144367274355,
  0.08507158180591605,
  0.07849762872395467,
  0.07164112578630137,
  0.06378286587358317,
  0.0552470114311219,
  0.0462945860424161,
  0.03712347442914019,
  0.02786842245114511,
  0.018601037106458284,
  0.009329786531283205,
  0,
  -0.008952971355287062,
  -0.017756530973588543,
  -0.027280669802972012,
  -0.03809382441738114,
  -0.05046287701663581,
  -0.06435315542643202,
  -0.07942843309834188,
  -0.09505092910981376,
  -0.110281308164172,
  -0.12398248438839508,
  -0.13610736529508502,
  -0.14737914491638404,
  -0.1583748269310998,
  -0.16950533281833757,
]
`;
